src/v/http/rest_client/rest_entity.h

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "http/client.h"
+#include "http/request_builder.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/sstring.hh>
+
+#include <absl/container/flat_hash_map.h>
+#include <boost/beast/http/verb.hpp>
+
+namespace http::rest_client {
+
+// A rest entity represents an API endpoint and provides methods for CRUD
+// operations on it. For example given a REST resource /person, the rest entity
+// person{} is expected to provide create, retrieve, update and delete
+// operations relative to that REST resource. This does not however represent a
+// single person object such as /person/1.
+class rest_entity {
+public:
+    // The root url is used as the parent path for a rest entity. So given a
+    // REST resource /api/v1/person, the root url should be specified as
+    // /api/v1.
+    explicit rest_entity(std::string_view root_url);
+
+    // The name of the resource which is appended to the root url to generate
+    // the full path. In the canonical person example, it should be "person".
+    // Along with a root url of "/api/v1" this will result in calls to the path
+    // "/api/v1/person"
+    virtual ss::sstring resource_name() const = 0;
+
+    using verb = boost::beast::http::verb;
+    virtual verb create_verb() const { return verb::post; }
+    virtual verb retrieve_verb() const { return verb::get; }
+    virtual verb update_verb() const { return verb::post; }
+    virtual verb delete_verb() const { return verb::delete_; }
+
+    virtual ~rest_entity() = default;
+
+    // Optional args which can be supplied with any request. Any optional
+    // headers or query params are applied to the HTTP request
+    using optional_headers
+      = std::optional<absl::flat_hash_map<std::string_view, std::string_view>>;
```

Comment:
Just curious, what are going to be the expected lifetimes of the items in the map? Are they always going to be static strings? Wondering if these might point to things like a topic/table name if we're guaranteed they'll always outlive the calls that use these

Reply:
The items are only needed until the call which accepts optional_headers ends, which is a non coroutine call in all cases. In the calls the string views are used to set fields on the request header or initialize map of `ss::sstring`

I used `string_view` here as the `boost::beast` http request accepts string_view directly whereas for `ss::sstring` there is a bit of conversion involved, the items in the map are used as input for the `boost::beast::http::request_header::set` call for headers or for initializing the `ss::sstring` values in the request builder query params map. 

I assume that since all the methods on request builder and rest entity are non async this should be safe, but if there are possible issues you can see with lifetimes we can switch to owned strings like `ss::sstring`.

Reply:
I would suggest that it's safer to move to `ss::sstring`. Reasoning:

1. the copy here probably doesn't matter, this isn't a perf critical path
2. most of these strings will fit within SSO anyways, and the allocation is probably more expensive anyways

Reply:
changed